### PR TITLE
external: add support for IPV6 ceph daemons

### PR DESCRIPTION
**Pull request by parth-gr**
updated some function which were strictly used for IPV4

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves # https://github.com/rook/rook/issues/11602

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: deploy/examples/create-external-cluster-resources.py [hunk 1]
```diff
@@ -762,21 +747,26 @@ def get_active_and_standby_mgrs(self):
...
+                monitoring_endpoint_ip, monitoring_endpoint_port
             )
             # collect all the 'stand-by' mgr ips
             mgr_ips = []
             for each_standby_mgr in standby_mgrs:
                 failed_ip = each_standby_mgr
-                mgr_ips.append(self._convert_hostname_to_ip(each_standby_mgr))
+                mgr_ips.append(
+                    self._convert_hostname_to_ip(
+                        each_standby_mgr, monitoring_endpoint_port
```
Comment (parth-gr) [inline deploy/examples/create-external-cluster-resources.py:758]: The port for standby_mgr would be also same, right?

Comment (travisn) [reply deploy/examples/create-external-cluster-resources.py:758]: Yes, i'd expect the port to be the same between active and standby mgr

**Comment by parth-gr** [comment]
How a user can input Ipv6 adrress,

Is it like, [2001:0000:3238:DFE1:0063:0000:0000:FEFB]:port or 2001:0000:3238:DFE1:0063:0000:0000:FEFB:port, or should we consider bioth ways?

**Comment by travisn** [comment]
> How a user can input Ipv6 adrress,
> 
> Is it like, [2001:0000:3238:DFE1:0063:0000:0000:FEFB]:port or 2001:0000:3238:DFE1:0063:0000:0000:FEFB:port, or should we consider both ways?

Let's just not try to validate the ipv6 format with our own parsing. There should be a library to validate the ipv6 address.

File: deploy/examples/create-external-cluster-resources.py [hunk 2]
```diff
@@ -540,35 +541,53 @@ def _check_conflicting_options(self):
...
+            endpoint_str_ip, port = endpoint_str.rsplit(":", 1)
         except ValueError:
-            raise ExecutionFailureException(
-                f"Not a proper endpoint: {endpoint_str}, <IPv4>:<PORT>, format is expected"
+            raise ExecutionFailureException(f"Not a proper endpoint: {endpoint_str}")
+
+        try:
+            if endpoint_str_ip[0] == "[":
+                # endpoint_str_ip = endpoint_str_ip[]
+            ip_type = (
```
Comment (subhamkrai) [inline deploy/examples/create-external-cluster-resources.py:553]:
```suggestion
                ip_type = (
```
indentation is needed?

File: deploy/examples/create-external-cluster-resources.py [hunk 3]
```diff
@@ -540,35 +541,53 @@ def _check_conflicting_options(self):
...
+                "IPv4" if type(ip_address(endpoint_str_ip)) is IPv4Address else "IPv6"
```
Comment (subhamkrai) [inline deploy/examples/create-external-cluster-resources.py:554]: We may also use `IPv6Address` library to check if it is ipv6 ip, currently IIUC, we are checking if it is not ip4 then make it ipv6?

**Review by subhamkrai — CHANGES_REQUESTED** [review]

File: deploy/examples/create-external-cluster-resources.py [hunk 4]
```diff
@@ -32,6 +32,7 @@
 if sys.version_info.major >= 3:
     py3k = True
     import urllib.parse
+    from ipaddress import ip_address, IPv4Address
```
Comment (subhamkrai) [inline deploy/examples/create-external-cluster-resources.py:35]: also, if this module is not there by default, we may need to document or mention

File: deploy/examples/create-external-cluster-resources.py (hunk 2 above)
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py:553]: Updated

File: deploy/examples/create-external-cluster-resources.py (hunk 4 above)
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources.py:35]: No it is the library

File: Documentation/CRDs/Cluster/external-cluster.md [hunk 5]
```diff
@@ -40,7 +40,7 @@ python3 create-external-cluster-resources.py --rbd-data-pool-name <pool_name> --
 - `--format bash`: The format of the output
 - `--rbd-data-pool-name`: The name of the RBD data pool
 - `--alias-rbd-data-pool-name`: Provides an alias for the  RBD data pool name, necessary if a special character is present in the pool name such as a period or underscore
-- `--rgw-endpoint`: (optional) The RADOS Gateway endpoint in the format `<IP>:<PORT>`
+- `--rgw-endpoint`: (optional) The RADOS Gateway endpoint in the format `<IP>:<PORT>`. Note: FQDN is also supported(in <FQDN>:<PORT> format)
```
[the remark this answers is not in the document]
Comment (parth-gr) [reply Documentation/CRDs/Cluster/external-cluster.md]: Updated

**Comment by parth-gr** [comment]
Testing:
1) just normal run 
```
[ceph: root@dev-parth ceph]# python3 a.py --rbd-data-pool-name rbd                                            
[{"name": "rook-ceph-mon-endpoints", "kind": "ConfigMap", "data": {"data": "dhcp46-22=[2620:52:0:462c:250:56ff:fe8f:dfca]:6789", "maxMonId": "0", "mapping": "{}"}}, {"name": "rook-ceph-mon", "kind": "Secret", "data": {"admin-secret": "admin-secret", "fsid": "5eb35978-0b71-11ee-b6f4-0050568f47e8", "mon-secret": "mon-secret"}}, {"name": "rook-ceph-operator-creds", "kind": "Secret", "data": {"userID": "client.healthchecker", "userKey": "AQDW/4pkQQcqERAATsNz0tENOuU8ZgJhAzBecA=="}}, {"name": "monitoring-endpoint", "kind": "CephCluster", "data": {"MonitoringEndpoint": "2620:52:0:462c:250:56ff:fe8f:47e8", "MonitoringPort": "9283"}}, {"name": "rook-csi-rbd-node", "kind": "Secret", "data": {"userID": "csi-rbd-node", "userKey": "AQDW/4pk3tyJERAA7HnYenpS0CQ5ix9Dect3uw=="}}, {"name": "rook-csi-rbd-provisioner", "kind": "Secret", "data": {"userID": "csi-rbd-provisioner", "userKey": "AQDW/4pkTbTjERAAB0CxgMg6TmHNhhhLxhY+hg=="}}, {"name": "rook-ceph-dashboard-link", "kind": "Secret", "data": {"userID": "ceph-dashboard-link", "userKey": "https://[2620:52:0:462c:250:56ff:fe8f:47e8]:8443/"}}, {"name": "ceph-rbd", "kind": "StorageClass", "data": {"pool": "rbd", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-rbd-node"}}]

```

2) Passing monitoring endpoint:
```
[ceph: root@dev-parth ceph]# python3 a.py --rbd-data-pool-name rbd --monitoring-endpoint 2620:52:0:462c:250:56ff:fe8f:47e8   
[{"name": "rook-ceph-mon-endpoints", "kind": "ConfigMap", "data": {"data": "dhcp46-22=[2620:52:0:462c:250:56ff:fe8f:dfca]:6789", "maxMonId": "0", "mapping": "{}"}}, {"name": "rook-ceph-mon", "kind": "Secret", "data": {"admin-secret": "admin-secret", "fsid": "5eb35978-0b71-11ee-b6f4-0050568f47e8", "mon-secret": "mon-secret"}}, {"name": "rook-ceph-operator-creds", "kind": "Secret", "data": {"userID": "client.healthchecker", "userKey": "AQDW/4pkQQcqERAATsNz0tENOuU8ZgJhAzBecA=="}}, {"name": "monitoring-endpoint", "kind": "CephCluster", "data": {"MonitoringEndpoint": "2620:52:0:462c:250:56ff:fe8f:47e8", "MonitoringPort": "9283"}}, {"name": "rook-csi-rbd-node", "kind": "Secret", "data": {"userID": "csi-rbd-node", "userKey": "AQDW/4pk3tyJERAA7HnYenpS0CQ5ix9Dect3uw=="}}, {"name": "rook-csi-rbd-provisioner", "kind": "Secret", "data": {"userID": "csi-rbd-provisioner", "userKey": "AQDW/4pkTbTjERAAB0CxgMg6TmHNhhhLxhY+hg=="}}, {"name": "rook-ceph-dashboard-link", "kind": "Secret", "data": {"userID": "ceph-dashboard-link", "userKey": "https://[2620:52:0:462c:250:56ff:fe8f:47e8]:8443/"}}, {"name": "ceph-rbd", "kind": "StorageClass", "data": {"pool": "rbd", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-rbd-node"}}]

```

3) passing monitoring endpoint hostname:
```
[ceph: root@dev-parth ceph]# python3 a.py --rbd-data-pool-name rbd --monitoring-endpoint dev-parth
[(<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('2620:52:0:462c:250:56ff:fe8f:47e8', 9283, 0, 0)), (<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('fe80::250:56ff:fe8f:47e8', 9283, 0, 0))] 2620:52:0:462c:250:56ff:fe8f:47e8
[{"name": "rook-ceph-mon-endpoints", "kind": "ConfigMap", "data": {"data": "dhcp46-22=[2620:52:0:462c:250:56ff:fe8f:dfca]:6789", "maxMonId": "0", "mapping": "{}"}}, {"name": "rook-ceph-mon", "kind": "Secret", "data": {"admin-secret": "admin-secret", "fsid": "5eb35978-0b71-11ee-b6f4-0050568f47e8", "mon-secret": "mon-secret"}}, {"name": "rook-ceph-operator-creds", "kind": "Secret", "data": {"userID": "client.healthchecker", "userKey": "AQDW/4pkQQcqERAATsNz0tENOuU8ZgJhAzBecA=="}}, {"name": "monitoring-endpoint", "kind": "CephCluster", "data": {"MonitoringEndpoint": "2620:52:0:462c:250:56ff:fe8f:47e8", "MonitoringPort": "9283"}}, {"name": "rook-csi-rbd-node", "kind": "Secret", "data": {"userID": "csi-rbd-node", "userKey": "AQDW/4pk3tyJERAA7HnYenpS0CQ5ix9Dect3uw=="}}, {"name": "rook-csi-rbd-provisioner", "kind": "Secret", "data": {"userID": "csi-rbd-provisioner", "userKey": "AQDW/4pkTbTjERAAB0CxgMg6TmHNhhhLxhY+hg=="}}, {"name": "rook-ceph-dashboard-link", "kind": "Secret", "data": {"userID": "ceph-dashboard-link", "userKey": "https://[2620:52:0:462c:250:56ff:fe8f:47e8]:8443/"}}, {"name": "ceph-rbd", "kind": "StorageClass", "data": {"pool": "rbd", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-rbd-node"}}]

```

**Comment by parth-gr** [comment]
having a mon name `"quorum_names":["dhcp46-22","dhcp46-23","dev-parth"` are fine?
Or we generally pick a,b,c?

**Comment by travisn** [comment]
> having a mon name `"quorum_names":["dhcp46-22","dhcp46-23","dev-parth"` are fine? Or we generally pick a,b,c?

In the external cluster, the names could be like those. In Rook, we just use a different convention with a,b,c, but Ceph allows any of those names.

**Review by travisn — APPROVED** [review]
Based on the testing, looks like we can go ahead with this.

**Review by subhamkrai — APPROVED** [review]

**Pull request merged**